tests/integration/security/reachability/reachability_test.go

```diff
@@ -44,280 +44,192 @@ import (
+						// Exclude calls to the naked app.
+						if opts.Target == naked {
+							return false
+						}
+
+						// Exclude calls to the headless TCP port.
+						if opts.Target == headless && opts.PortName == "tcp" {
+							return false
+						}
+						return true
+					},
+					expectSuccess: func(src echo.Instance, opts echo.CallOptions) bool {
+						return true
 					},
 				},
 				{
 					configFile: "global-mtls-off.yaml",
 					namespace:  systemNS,
-					subTests: []connection.Checker{
-						{
-							From: a,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
+					include: func(src echo.Instance, opts echo.CallOptions) bool {
+						// Exclude calls to the headless TCP port.
+						if opts.Target == headless && opts.PortName == "tcp" {
+							return false
+						}
+						return true
+					},
+					expectSuccess: func(src echo.Instance, opts echo.CallOptions) bool {
+						return true
 					},
 				},
 				{
 					configFile: "single-port-mtls-on.yaml",
 					namespace:  ns,
-					subTests: []connection.Checker{
-						{
-							From: a,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: false,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: false,
-						},
-						{
-							From: a,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "tcp",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "tcp",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
+					include: func(src echo.Instance, opts echo.CallOptions) bool {
+						// Include all tests that target app B, which has the single-port config.
+						return opts.Target == b
+					},
+					expectSuccess: func(src echo.Instance, opts echo.CallOptions) bool {
+						return opts.PortName != "http"
 					},
 				},
 			}
 
 			for _, c := range testCases {
 				testName := strings.TrimSuffix(c.configFile, filepath.Ext(c.configFile))
-				t.Run(testName, func(t *testing.T) {
+				ctx.NewSubTest(testName).
+					Run(func(ctx framework.TestContext) {
+						// Apply the policy.
+						policyYAML := file.AsStringOrFail(ctx, filepath.Join("testdata", c.configFile))
+						g.ApplyConfigOrFail(ctx, c.namespace, policyYAML)
+						_ = closer.NewOrFail(ctx, ctx, func() error {
+							return g.DeleteConfig(ctx, c.namespace.Name(), policyYAML)
```

Comment:
Would it make sense to change the contract of ApplyConfig to tie the lifecycle of the applied config to ctx? What you're doing here is, in effect , taken care of automatically. (We're already passing ctx in-lieu of T. We can expand its responsibility).

I am worried about the "closer approach like this, which is an escape hatch that is easy to abuse. Also, If we're going to do this, it maybe simpler to just add a method on TestContext, instead of exposing it as a component.

Reply:
In general, I think that would be fine. However, I'm worried about the nature of apply ... if it's used as a resource update, rather than creation, the opposite would not be to delete it, but rather revert it.  This is more difficult to handle automatically.

Reply:
ok, then. How about simplifying this to a lambda passed to the TestContext, to be called upon .Done()?

Reply:
done.